store/localstore/compactor.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package localstore
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/localstore/engine"
+	"github.com/pingcap/tidb/util/bytes"
+)
+
+var _ kv.Compactor = (*localstoreCompactor)(nil)
+
+const (
+	deleteWorkerSize = 3
+)
+
+var localCompactorDefaultPolicy = kv.CompactorPolicy{
+	SafeTime:        20 * 1000, // in ms
+	TriggerInterval: 1 * time.Second,
+	BatchDeleteSize: 100,
+}
+
+type localstoreCompactor struct {
+	mu         sync.Mutex
+	recentKeys map[string]struct{}
+	stopChan   chan struct{}
+	delChan    chan kv.EncodedKey
+	ticker     *time.Ticker
+	db         engine.DB
+	policy     kv.CompactorPolicy
+}
+
+func (gc *localstoreCompactor) OnSet(k kv.Key) {
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	gc.recentKeys[string(k)] = struct{}{}
+}
+
+func (gc *localstoreCompactor) OnGet(k kv.Key) {
+	// Do nothing now.
+}
+
+func (gc *localstoreCompactor) OnDelete(k kv.Key) {
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	gc.recentKeys[string(k)] = struct{}{}
+}
+
+func (gc *localstoreCompactor) getAllVersions(k kv.Key) ([]kv.EncodedKey, error) {
+	startKey := MvccEncodeVersionKey(k, kv.MaxVersion)
+	endKey := MvccEncodeVersionKey(k, kv.MinVersion)
+
+	it, err := gc.db.Seek(startKey)
```

Comment:
refactor these code involved redesign dbSnapshot's buffered iterator.  I'll open another PR to do this.